server/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java

```diff
@@ -181,6 +183,43 @@ private static Query newObjectFieldExistsQuery(QueryShardContext context, String
         return new ConstantScoreQuery(booleanQuery.build());
     }
 
+    /**
+     * Helper method to get field mapped to this fieldPattern
+     * @return return collection of fields if exists else return empty.
+     */
+    private static Collection<String> getMappedField(QueryShardContext context, String fieldPattern) {
+        final FieldNamesFieldMapper.FieldNamesFieldType fieldNamesFieldType = (FieldNamesFieldMapper.FieldNamesFieldType) context
+            .getMapperService().fieldType(FieldNamesFieldMapper.NAME);
+
+        if (fieldNamesFieldType == null) {
+            // can only happen when no types exist, so no docs exist either
+            return Collections.emptySet();
+        }
+
+        final Collection<String> fields;
+        if (context.getObjectMapper(fieldPattern) != null) {
+            // the _field_names field also indexes objects, so we don't have to
+            // do any more work to support exists queries on whole objects
+            fields = Collections.singleton(fieldPattern);
+        } else {
+            fields = context.simpleMatchToIndexNames(fieldPattern);
+        }
+
+        if (fields.size() == 1) {
+            String field = fields.iterator().next();
+            MappedFieldType fieldType = context.getMapperService().fieldType(field);
+            if (fieldType == null) {
+                // The field does not exist as a leaf but could be an object so
+                // check for an object mapper
+                if (context.getObjectMapper(field) == null) {
+                    return Collections.emptySet();
+                }
+            }
+        }
```

Comment:
I believe that this entire if statement can be removed without affecting correctness. It is not possible to have a field that is neither and object nor a field at this stage.

Reply:
But in this, `if` statement we are checking the field type. 
` MappedFieldType fieldType = context.getMapperService().fieldType(field);`

Reply:
Thanks I had the wrong assumption that simpleMatchToIndexName would return an empty set if the field is not mapped.